docs/architecture/adr-105-refactor-mempool-senders.md

```diff
@@ -0,0 +1,141 @@
+# ADR 105: Refactor list of senders in mempool
+
+## Changelog
+
+- 2023-06-26: Initial draft (@hvanz)
+
+## Status
+
+Proposed
+
+## Context
+
+Before adding a transaction to the mempool, we need to send a `CheckTx` message
+to the application for validating the transaction. There are [two variants][CheckTxType] of
+this message, distinguished by the value in message field `type`:
+- `CheckTxType_New` is for transactions that need to be validated before adding
+it to the mempool for the first time.
+- `CheckTxType_Recheck` is for transactions that are in the mempool and need to
+be re-validated after committing a block and advancing to the next height.
+
+The mempool communicates with the ABCI server (that is, the application) by
+sending `abci.Request`s through the proxy `AppConnMempool`. The proxy provides a
+callback mechanism for handling `abci.Response`s. The current mempool
+implemention `CListMempool` (also called v0) utilizes this mechanism for
+`Recheck` transactions but not for `New` transactions. Instead `New`
+transactions require an ad-hoc mechanism for each request. 
+
+The reason behind this difference is that for `New` transactions we need to
+record the ID of the peer that sent the transaction. However, this information
+is not included in `RequestCheckTx` messages. Recording the sender's ID is
+necessary for the transaction propagation protocol, which uses the recorded list
+of senders to prevent sending the transaction back to these peers, thus avoiding
+duplicated messages. More importantly, this mechanism serves as the only means
+to stop propagating transactions.
```

Comment:
Are we sure we'll still need this info, in this way, by the end of Q3?

Reply:
I'm mostly certain that we will still use a mechanism like this. Now we keep track of the senders of each transaction. Probably we will expand this set to include any node that is known to have the transactions. In any case, this ADR can be superseded by another ADR.